tezos/new_context/src/working_tree/working_tree.rs

```diff
@@ -558,56 +488,46 @@ impl WorkingTree {
 
     /// Returns a new version of the WorkingTree with the tree replaced
     pub fn with_new_root(&self, tree: Tree) -> Self {
-        let staged_cache = Rc::clone(&self.staged_cache);
+        let index = self.index.clone();
         let stats = self.stats.clone();
-        Self {
-            tree,
-            staged_cache,
-            stats,
-            ..*self
-        }
+        Self { tree, stats, index }
     }
 
-    /// Set key/val to the staging area.
+    /// Set key/val to the working tree.
     pub fn set(&self, key: &ContextKey, value: ContextValue) -> Result<Self, MerkleError> {
-        //let stat_updater = StatUpdater::new(MerkleStorageAction::Set, Some(key));
-        let new_root_hash = &self._set(&self.tree, key, value)?;
-        //self.set_stage_root(&self.get_tree(new_root_hash)?, new_tree_id);
-        let tree = self.get_tree(new_root_hash)?;
-        //stat_updater.update_execution_stats(&mut self.stats);
+        // let stat_updater = StatUpdater::new(MerkleStorageAction::Set, Some(key));
+
+        let entry = &self._set(key, value)?;
+        let tree = self.get_tree(entry)?.clone();
+
+        // stat_updater.update_execution_stats(&mut self.stats);
         Ok(self.with_new_root(tree))
     }
 
-    fn _set(
-        &self,
-        root: &Tree,
-        key: &ContextKey,
-        value: ContextValue,
-    ) -> Result<Arc<EntryHash>, MerkleError> {
-        let blob_hash = Arc::new(hash_blob(&value)?);
-        self.put_to_staging_area(blob_hash.clone(), Entry::Blob(value.clone()));
+    fn _set(&self, key: &ContextKey, value: ContextValue) -> Result<Entry, MerkleError> {
+        let entry = RefCell::new(Some(Entry::Blob(value)));
         let new_node = Node {
-            entry_hash: blob_hash,
+            entry,
+            entry_hash: RefCell::new(None),
             node_kind: NodeKind::Leaf,
         };
-        self.compute_new_root_with_change(root, &key, Some(new_node))
+        self.compute_new_root_with_change(&key, Some(new_node))
     }
 
     /// Delete an item from the staging area.
     pub fn delete(&self, key: &ContextKey) -> Result<Self, MerkleError> {
         //let stat_updater = StatUpdater::new(MerkleStorageAction::Delete, Some(key));
-        let new_root_hash = &self._delete(&self.tree, key)?;
-        //self.set_stage_root(&self.get_tree(new_root_hash)?, new_tree_id);
-        let tree = self.get_tree(new_root_hash)?;
+        let new_root_hash = &self._delete(key)?;
```

Comment:
`new_root_hash` -> `new_root_entry` because of the change in `_delete`'s return type.